{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import bq_helper\n",
    "from bq_helper import BigQueryHelper\n",
    "# https://www.kaggle.com/sohier/introduction-to-the-bq-helper-package\n",
    "genomes = bq_helper.BigQueryHelper(active_project=\"bigquery-public-data\",\n",
    "                                   dataset_name=\"genomics_cannabis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MNPR01_201703',\n",
       " 'MNPR01_reference_201703',\n",
       " 'MNPR01_transcriptome_201703',\n",
       " 'cs10_gff',\n",
       " 'cs3k_project_info',\n",
       " 'cs3k_vcf_cs10_dv090',\n",
       " 'sample_info_201703']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bq_assistant = BigQueryHelper(\"bigquery-public-data\", \"genomics_cannabis\")\n",
    "bq_assistant.list_tables()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\sharm\\AppData\\Roaming\\Python\\Python311\\site-packages\\google\\cloud\\bigquery\\table.py:1727: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>alternate_bases</th>\n",
       "      <th>quality</th>\n",
       "      <th>type</th>\n",
       "      <th>AB</th>\n",
       "      <th>ABP</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>AO</th>\n",
       "      <th>...</th>\n",
       "      <th>RPP</th>\n",
       "      <th>RPPR</th>\n",
       "      <th>RPR</th>\n",
       "      <th>RUN</th>\n",
       "      <th>SAF</th>\n",
       "      <th>SAP</th>\n",
       "      <th>SAR</th>\n",
       "      <th>SRF</th>\n",
       "      <th>SRP</th>\n",
       "      <th>SRR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTI1NzF8Z2J8TU5QUj...</td>\n",
       "      <td>[CTA]</td>\n",
       "      <td>23.39620</td>\n",
       "      <td>[snp]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTEyODV8Z2J8TU5QUj...</td>\n",
       "      <td>[CGTAA]</td>\n",
       "      <td>7.52241</td>\n",
       "      <td>[complex]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTY5MTJ8Z2J8TU5QUj...</td>\n",
       "      <td>[CCTT]</td>\n",
       "      <td>25.30020</td>\n",
       "      <td>[snp]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTY5OTl8Z2J8TU5QUj...</td>\n",
       "      <td>[TATGG]</td>\n",
       "      <td>6.16805</td>\n",
       "      <td>[complex]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0ODY2OTF8Z2J8TU5QUj...</td>\n",
       "      <td>[ATTTTTTTAA]</td>\n",
       "      <td>3.08123</td>\n",
       "      <td>[complex]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTQyOTF8Z2J8TU5QUj...</td>\n",
       "      <td>[GAG]</td>\n",
       "      <td>23.56570</td>\n",
       "      <td>[snp]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTUwNTV8Z2J8TU5QUj...</td>\n",
       "      <td>[GC]</td>\n",
       "      <td>27.00690</td>\n",
       "      <td>[del]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTUyMDF8Z2J8TU5QUj...</td>\n",
       "      <td>[GATAAA]</td>\n",
       "      <td>14.01670</td>\n",
       "      <td>[mnp]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0OTU1NTV8Z2J8TU5QUj...</td>\n",
       "      <td>[TTTTGC]</td>\n",
       "      <td>12.62600</td>\n",
       "      <td>[mnp]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CKXG8eKP9qOf8wESIGdpfDEwOTg0ODgxOTZ8Z2J8TU5QUj...</td>\n",
       "      <td>[AGT]</td>\n",
       "      <td>9.12466</td>\n",
       "      <td>[snp]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[0.0]</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[1.0]</td>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[5.18177]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          variant_id alternate_bases  \\\n",
       "0  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTI1NzF8Z2J8TU5QUj...           [CTA]   \n",
       "1  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTEyODV8Z2J8TU5QUj...         [CGTAA]   \n",
       "2  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTY5MTJ8Z2J8TU5QUj...          [CCTT]   \n",
       "3  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTY5OTl8Z2J8TU5QUj...         [TATGG]   \n",
       "4  CKXG8eKP9qOf8wESIGdpfDEwOTg0ODY2OTF8Z2J8TU5QUj...    [ATTTTTTTAA]   \n",
       "5  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTQyOTF8Z2J8TU5QUj...           [GAG]   \n",
       "6  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTUwNTV8Z2J8TU5QUj...            [GC]   \n",
       "7  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTUyMDF8Z2J8TU5QUj...        [GATAAA]   \n",
       "8  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTU1NTV8Z2J8TU5QUj...        [TTTTGC]   \n",
       "9  CKXG8eKP9qOf8wESIGdpfDEwOTg0ODgxOTZ8Z2J8TU5QUj...           [AGT]   \n",
       "\n",
       "    quality       type     AB    ABP   AC     AF  AN   AO  ...        RPP  \\\n",
       "0  23.39620      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "1   7.52241  [complex]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "2  25.30020      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "3   6.16805  [complex]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "4   3.08123  [complex]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "5  23.56570      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "6  27.00690      [del]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "7  14.01670      [mnp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "8  12.62600      [mnp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "9   9.12466      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   \n",
       "\n",
       "   RPPR  RPR  RUN  SAF        SAP  SAR  SRF  SRP  SRR  \n",
       "0   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "1   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "2   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "3   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "4   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "5   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "6   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "7   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "8   0.0  [0]  [1]  [1]  [5.18177]  [0]    0  0.0    0  \n",
       "9   0.0  [1]  [1]  [0]  [5.18177]  [1]    0  0.0    0  \n",
       "\n",
       "[10 rows x 39 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"SELECT\n",
    "  variant_id, alternate_bases, quality, type, \n",
    "  AB, ABP, AC, AF, AN, AO, DP, DPB, DPRA, EPP, \n",
    "  EPPR, GTI, MQM, MQMR, NS, NUMALT, ODDS, PAO,\n",
    "  PQA, PQR, PRO, QA, QR, RO, RPL, RPP, RPPR, \n",
    "  RPR, RUN, SAF, SAP, SAR, SRF, SRP, SRR\n",
    "FROM\n",
    "  `bigquery-public-data.genomics_cannabis.MNPR01_201703` v\n",
    "limit 1000000;\"\"\"\n",
    "response = genomes.query_to_pandas_safe(query, max_gb_scanned=100)\n",
    "response.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I shall run regression to determine my DNA samples based on proteins. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "response[\"DNA_A\"] = response[\"alternate_bases\"].str[0].str.count('A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "response[\"DNA_C\"] = response[\"alternate_bases\"].str[0].str.count('C')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "response[\"DNA_G\"] = response[\"alternate_bases\"].str[0].str.count('G')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "response[\"DNA_T\"] = response[\"alternate_bases\"].str[0].str.count('T')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I shall adjust the proteins for easy analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AB\n",
      "ABP\n",
      "AC\n",
      "AF\n",
      "AO\n",
      "DPRA\n",
      "EPP\n",
      "MQM\n",
      "PAO\n",
      "PQA\n",
      "QA\n",
      "RPL\n",
      "RPP\n",
      "RPR\n",
      "RUN\n",
      "SAF\n",
      "SAP\n",
      "SAR\n"
     ]
    }
   ],
   "source": [
    "proteins = [\"AB\", \"ABP\", \"AC\", \"AF\", \"AO\", \"DPRA\", \"EPP\", \"MQM\", \"PAO\",\n",
    "            \"PQA\", \"QA\", \"RPL\", \"RPP\", \n",
    "            \"RPR\", \"RUN\", \"SAF\", \"SAP\", \"SAR\", \n",
    "]\n",
    "\n",
    "for protein in proteins:\n",
    "    print(protein)\n",
    "    response[protein] = response[protein].str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "response[\"type\"] = response[\"type\"].str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "snp        699662\n",
       "complex    158031\n",
       "mnp         57898\n",
       "ins         51333\n",
       "del         33076\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response[\"type\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```type``` is a categorical variable, so I shall need to dummify it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex</th>\n",
       "      <th>del</th>\n",
       "      <th>ins</th>\n",
       "      <th>mnp</th>\n",
       "      <th>snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        complex  del  ins  mnp  snp\n",
       "0             0    0    0    0    1\n",
       "1             1    0    0    0    0\n",
       "2             0    0    0    0    1\n",
       "3             1    0    0    0    0\n",
       "4             1    0    0    0    0\n",
       "...         ...  ...  ...  ...  ...\n",
       "999995        0    0    0    0    1\n",
       "999996        1    0    0    0    0\n",
       "999997        0    0    0    0    1\n",
       "999998        1    0    0    0    0\n",
       "999999        0    0    0    0    1\n",
       "\n",
       "[1000000 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dummies = pd.get_dummies(response[\"type\"], dtype=\"int\")\n",
    "dummies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I shall define what I am regressing upon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "y = response[[\"DNA_A\", \"DNA_C\", \"DNA_G\", \"DNA_T\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = pd.concat([\n",
    "    response[[\"AB\", \"ABP\", \"AC\", \"AF\", \"AN\", \"AO\", \"DP\", \"DPB\", \"DPRA\", \"EPP\", \n",
    "              \"EPPR\", \"GTI\", \"MQM\", \"MQMR\", \"NS\", \"NUMALT\", \"ODDS\", \"PAO\",\n",
    "              \"PQA\", \"PQR\", \"PRO\", \"QA\", \"QR\", \"RO\", \"RPL\", \"RPP\", \"RPPR\", \n",
    "              \"RPR\", \"RUN\", \"SAF\", \"SAP\", \"SAR\", \"SRF\", \"SRP\", \"SRR\"\n",
    "    ]], \n",
    "    dummies\n",
    "], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LinearRegression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train score 0.17432333633747393, test score 0.15232295290835307\n"
     ]
    }
   ],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.33, random_state=1337)\n",
    "LR = LinearRegression()\n",
    "LR.fit(X_train, y_train)\n",
    "train_score = LR.score(X_train, y_train)\n",
    "test_score = LR.score(X_test, y_test)\n",
    "print(f'Train score {train_score}, test score {test_score}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I've achieved a test score of 15.2%. This is unsatisfactory, I will need a different method. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
